crates/bdk/src/wallet/mod.rs

```diff
@@ -964,176 +1192,200 @@ impl<D> Wallet<D> {
-                        let satisfaction_weight =
-                            serialize(&txin.script_sig).len() * 4 + serialize(&txin.witness).len();
-                        WeightedUtxo {
-                            satisfaction_weight,
-                            utxo: Utxo::Foreign {
-                                outpoint: txin.previous_output,
-                                psbt_input: Box::new(psbt::Input {
-                                    witness_utxo: Some(txout.clone()),
-                                    non_witness_utxo: Some(prev_tx.clone()),
-                                    ..Default::default()
-                                }),
-                            },
-                        }
-                    }
-                };
+        let txout_index = &mut self.keychain_tracker.txout_index;
+        for (k, i) in to_unamrk {
+            txout_index.unmark_used(&k, i);
+        }
+    }
 
-                Ok(weighted_utxo)
+    /// Return the list of unspent outputs of this wallet
+    pub fn list_unspent(&self) -> Vec<(LocalUtxo, K)> {
+        self.keychain_tracker
+            .full_utxos()
+            .map(|((keychain, derivation_index), utxo)| {
+                (
+                    LocalUtxo {
+                        outpoint: utxo.outpoint,
+                        txout: utxo.txout,
+                        is_spent: false,
+                        derivation_index: *derivation_index,
+                        confirmation_time: utxo.chain_position,
+                    },
+                    keychain.clone(),
+                )
             })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        if tx.output.len() > 1 {
-            let mut change_index = None;
-            for (index, txout) in tx.output.iter().enumerate() {
-                let change_type = self.map_keychain(KeychainKind::Internal);
-                match txout_index.index_of_spk(&txout.script_pubkey) {
-                    Some(&(keychain, _)) if keychain == change_type => change_index = Some(index),
-                    _ => {}
-                }
-            }
-
-            if let Some(change_index) = change_index {
-                tx.output.remove(change_index);
-            }
-        }
+            .collect()
+    }
 
-        let params = TxParams {
-            // TODO: figure out what rbf option should be?
-            version: Some(tx_builder::Version(tx.version)),
-            recipients: tx
-                .output
-                .into_iter()
-                .map(|txout| (txout.script_pubkey, txout.value))
-                .collect(),
-            utxos: original_utxos,
-            bumping_fee: Some(tx_builder::PreviousFee {
-                absolute: fee,
-                rate: feerate.as_sat_per_vb(),
-            }),
-            ..Default::default()
-        };
+    /// Iterator over all keychains in this wallet
+    pub fn keychanins(&self) -> &BTreeMap<K, ExtendedDescriptor> {
```

Comment:
Nits: I think we are suppose to have `keychains` here instead of `keychanins`.

Reply:
Totally..